src/macros.jl

```diff
@@ -237,7 +237,126 @@ macro grad(expr)
+
+The `@grad_from_chainrules` macro provides a way to import
+adjoints(rrule) defined in ChainRules to ReverseDiff. One must provide
+a method signature to import the corresponding `rrule`. In the
+provided method signature, one should replace the types of arguments
+to which one wants to take derivatives with respect with
+`ReverseDiff.TrackedReal` and `ReverseDiff.TrackedArray`
+respectively. For example, we can import `rrule` of `f(::Real,
+::Array)` like below:
+
+
+```
+ReverseDiff.@grad_from_chainrules f(x::TrackedReal, ::TrackedArray)
+ReverseDiff.@grad_from_chainrules f(x::TrackedReal, ::Array)
+ReverseDiff.@grad_from_chainrules f(x::Real, ::TrackedArray)
+```
+
+"""
+macro grad_from_chainrules(fcall)
+    @gensym tp output_value output back closure cls_args cls_kwargs
+    f, fwd_def = funcall_fwd_to_rule(fcall)
+
+    return quote
+        $fwd_def
+        function $ReverseDiff.track(::typeof($f), args...; kwargs...)
+            $tp = $ReverseDiff.tape(args...)
+            $output_value, $back = ChainRules.rrule($f, map(ReverseDiff.value, args)...; kwargs...)
+            $output = $ReverseDiff.track($output_value, $tp)
+            $closure($cls_args...; $cls_kwargs...) = ChainRules.rrule($f, map(ReverseDiff.value, $cls_args)...; $cls_kwargs...)
+            $ReverseDiff.record!(
+                $tp,
+                $ReverseDiff.SpecialInstruction,
+                $f,
+                args,
+                $output,
+                ($back, $closure, kwargs),
+            )
+            return $output
+        end
+
+        if !hasmethod(
+            $ReverseDiff.special_reverse_exec!,
+            Tuple{$ReverseDiff.SpecialInstruction{typeof($f)}},
+        )
+            @noinline function $ReverseDiff.special_reverse_exec!(instruction::$ReverseDiff.SpecialInstruction{typeof($f)})
+                output = instruction.output
+                input = instruction.input
+                back = instruction.cache[1]
+                back_output = back($ReverseDiff.deriv(output))
+                input_derivs = back_output[2:end]
+                @assert input_derivs isa Tuple
+                $ReverseDiff._add_to_deriv!.(input, input_derivs)
+                $ReverseDiff.unseed!(output)
+                return nothing
+            end
+        end
+
+        if !hasmethod(
+            $ReverseDiff.special_forward_exec!,
+            Tuple{$ReverseDiff.SpecialInstruction{typeof($f)}},
+        )
+            @noinline function $ReverseDiff.special_forward_exec!(instruction::$ReverseDiff.SpecialInstruction{typeof($f)})
+                output, input = instruction.output, instruction.input
+                $ReverseDiff.pull_value!.(input)
+                pullback = instruction.cache[2]
+                kwargs = instruction.cache[3]
+                out_value = pullback(input...; kwargs...)[1]
+                $ReverseDiff.value!(output, out_value)
+                return nothing
+            end
+        end
+    end |> esc
```

Comment:
Can we not escape the whole expression but only the parts that have to be escaped (like the function name and user-provided types) and let macro hygiene take care of the rest? Similar to e.g. the macros in ChainRulesCore such as https://github.com/JuliaDiff/ChainRulesCore.jl/blob/47389f5d4748209a88c5eb72a7cdf50cdd8bfa2d/src/rule_definition_tools.jl#L246-L254.

Reply:
There are too many symbols to preserve(`esc`) in this piece of code, using `gensym` to generate new symbols and keep the macro hygiene in this way makes it more readable, I think it was also because of this, the macro `@grad` was written in this style?

Reply:
I just think nobody cared when the other macro was written and/or nobody wanted to deal with macro hygiene more carefully. Also initially the helpers in ChainRulesCore just escaped everything and were only changed a bit later.

Without `esc` you could get rid of most of the interpolations (`$ReverseDiff` etc.), which would even improve readability it seems. We mainly have to preserve (`esc`) `f` and the types in the user-provided function signature (which we have to handle anyway in `_make_fwd_args`). We could even avoid escaping the gensymed variables (and remove them completely) if we use custom argument names such as `args1` etc in the function signature (so that it is ensure there are no clashes with variable names in the function body).